Source/WebCore/dom/CustomStateSet.cpp

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "CustomStateSet.h"
+#include "Element.h"
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(CustomStateSet);
+
+bool CustomStateSet::addToSetLike(String state)
+{
+    const AtomString state_atom = AtomString(state);
+    bool removed = m_states.removeFirst(&state_atom);
+    m_states.append(&state_atom);
+    Element* element = m_element.get();
+    if (element && !removed)
+        element->invalidateStyleForCustomStateChange();
```

Comment:
I don't think this would actually invalidate. The PseudoClassInvalidationChange needs to be in the same scope as when the matching changes. You'll want something among those lines:
```suggestion
    m_states.append(&state_atom);
    Element* element = m_element.get();
    std::optional<Style::PseudoClassChangeInvalidation> styleInvalidation;
    if (element && !removed)
        styleInvalidation.emplace(m_element, CSSSelector::PseudoClassType::State, Style::PseudoClassChangeInvalidation::AnyValue);
```

Reply:
Also it would be good to have invalidation WPT for this. Simple invalidation and also invalidation with `:has()`.

See:
LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/defined-in-has.html
LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/defined.html